internal/controller/stages/regular_stages.go

```diff
@@ -322,6 +322,11 @@ func (r *RegularStageReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, nil
 	}
 
+	if stage.GetLabels()[kargoapi.LabelKeyShard] != r.cfg.ShardName {
+		logger.Debug("ignoring Stage because it is is not assigned to this shard")
+		return ctrl.Result{}, nil
+	}
```

Comment:
As we discussed offline, #4708 changes the logic required here, but luckily also introduces a new predicate, which you can probably just use directly right here.

fwiw, I would probably make that predicate a field of the reconciler and initialize it just once when the reconciler is initialized.